### PR TITLE
Sofia/slosilo key for hosts and users

### PR DESCRIPTION
**Pull request by sofiadimant1**
### Desired Outcome

*Please describe the desired outcome for this PR.  Said another way, what was
the original request that resulted in these code changes?  Feel free to copy
this information from the connected issue.*

### Implemented Changes

*Describe how the desired outcome above has been achieved with this PR. In
particular, consider:*

- _What's changed? Why were these changes made?_
- _How should the reviewer approach this PR, especially if manual tests are required?_
- _Are there relevant screenshots you can add to the PR description?_

### Connected Issue/Story

Resolves #[relevant GitHub issue(s), e.g. 76]

CyberArk internal issue ID: [insert issue ID]

### Definition of Done
*At least 1 todo must be completed in the sections below for the PR to be
merged.*

#### Changelog

- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ ] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

### CODE REVIEW

File: app/models/account.rb [hunk 1]
```diff
@@ -40,34 +49,38 @@ def create(id, owner_id = nil)
...
         $1
       end.delete_if do |account|
         account == "!"
       end
     end
   end
 
-  def token_key
-    Slosilo["authn:#{id}"]
+  def token_key_host
```
Comment (jvanderhoof) [inline app/models/account.rb:65]: What do you think about combining these two methods with the above slosilo id methods to make them a bit more generic, something like:

```ruby
def token_key(role)
  Slosilo[token_id(role)]
end

def token_id(role)
  "authn:#{account}:#{role}"
end
```

File: app/models/account.rb [hunk 2]
```diff
@@ -17,15 +17,24 @@ def find_or_create_accounts_resource
...
+      "authn:#{account}:host"
+    end
+
+    def user_slosilo_id(account)
+      "authn:#{account}:user"
+    end
+
     def create(id, owner_id = nil)
-      raise Exceptions::RecordExists.new("account", id) if Slosilo["authn:#{id}"]
+      raise Exceptions::RecordExists.new("account", id) if Slosilo[host_slosilo_id(id)] or Slosilo[user_slosilo_id(id)]
```
Comment (jvanderhoof) [inline app/models/account.rb:29]: Alternatively, you can use the `token_key...` methods below (which return the key if present:

```ruby
raise Exceptions::RecordExists.new("account", id) if token_key('user') || token_key('host')
```

Comment (jvanderhoof) [reply app/models/account.rb:29]: You likely know this, but this line will raise if either host or user keys are missing. It the migration the only bit ensuring both keys are present?

Comment (sofiadimant1) [reply app/models/account.rb]: what do you mean? there is a rspec test to test it

**Review by jeniaSakirko — APPROVED** [review]

**Pull request merged**